pkg/awsutils/awsutils.go

```diff
@@ -992,33 +1173,79 @@ func (cache *EC2InstanceMetadataCache) createENI(sg []*string, eniCfgSubnet stri
 			return networkInterfaceID, nil
 		}
 	} else {
-		if cache.useSubnetDiscovery && !cache.v6Enabled {
+		if cache.useSubnetDiscovery {
 			subnetResult, vpcErr := cache.getVpcSubnets()
 			if vpcErr != nil {
 				log.Warnf("Failed to call ec2:DescribeSubnets: %v", vpcErr)
 				log.Info("Defaulting to same subnet as the primary interface for the new ENI")
+
+				// Even in fallback, check if primary subnet is excluded
+				excluded, checkErr := cache.isPrimarySubnetExcluded()
+				if checkErr != nil {
+					log.Warnf("Failed to check if primary subnet is excluded: %v. Proceeding with ENI creation attempt.", checkErr)
+				} else if excluded {
+					// Primary subnet is explicitly excluded
+					err = errors.New("primary subnet is tagged with kubernetes.io/role/cni=0 - no valid subnets available for ENI creation")
+					log.Error(err.Error())
+					return "", err
+				}
+
 				networkInterfaceID, err = cache.tryCreateNetworkInterface(input)
 				if err == nil {
 					return networkInterfaceID, nil
 				}
 			} else {
+				validSubnetsFound := false
 				for _, subnet := range subnetResult {
-					if *subnet.SubnetId != cache.subnetID {
-						if !validTag(subnet) {
-							continue
+					// Check tag for all subnets including primary
+					isPrimarySubnet := *subnet.SubnetId == cache.subnetID
+					if !validTag(subnet, isPrimarySubnet) {
+						// Log when primary subnet is excluded
+						if isPrimarySubnet {
+							log.Infof("Primary subnet %s is excluded from ENI creation", cache.subnetID)
 						}
+						continue
+					}
+					validSubnetsFound = true
+					// If this is a secondary subnet and we have custom security groups, use those instead
+					// We already determined isPrimarySubnet above, just reuse the variable
+					if !isPrimarySubnet && len(cache.customSecurityGroups.SortedList()) > 0 {
+						log.Infof("Using custom security groups for ENI in secondary subnet %s", *subnet.SubnetId)
+						input.Groups = cache.customSecurityGroups.SortedList()
+					} else if !isPrimarySubnet {
+						// Secondary subnet but no custom security groups available - use primary SGs as fallback
+						log.Infof("No custom security groups available, using primary security groups for ENI in secondary subnet %s", *subnet.SubnetId)
 					}
-					log.Infof("Creating ENI with security groups: %v in subnet: %s", input.Groups, aws.ToString(input.SubnetId))
+					log.Infof("Creating ENI with security groups: %v in subnet: %s", input.Groups, aws.ToString(subnet.SubnetId))
 
 					input.SubnetId = subnet.SubnetId
 					networkInterfaceID, err = cache.tryCreateNetworkInterface(input)
 					if err == nil {
 						return networkInterfaceID, nil
 					}
 				}
+
+				// If no valid subnets found, return appropriate error
+				if !validSubnetsFound {
+					err = errors.New("no valid subnets available for ENI creation - all subnets are either not tagged or tagged with kubernetes.io/role/cni=0")
```

Comment:
Can you check the similar pattern to use one line return instead of three lines.